C/dag.c

```diff
@@ -268,3 +325,94 @@ void forceJets(dag_node* dag, const analyses* analysis, const size_t len, JET_FL
     }
   }
 }
+
+/* This function fills in the 'WITNESS' nodes of a 'dag' with the data from 'witness'.
+ * For each 'WITNESS' : A |- B expression in 'dag', the bits from the 'witness' bitstring are decoded in turn
+ * to construct a compact representation of a witness value of type B.
+ * This funciton only returns 'true' when exactly 'witness.len' bits are consumed by all the 'dag's witness values.
+ *
+ * Note: the 'witness' value is passed by copy because the implementation manipulates a local copy of the structure.
+ *
+ * Precondition: dag_node dag[len] and 'dag' without witness data and is well-typed with 'type_dag';
+ *               witness is a valid bitstring;
+ *
+ * Postcondition: dag_node dag[len] and 'dag' has witness data and is well-typed with 'type_dag'
+ *                  when the result is 'true';
+ */
+bool fillWitnessData(dag_node* dag, type* type_dag, const size_t len, bitstring witness) {
+  for (size_t i = 0; i < len; ++i) {
+    if (WITNESS == dag[i].tag) {
+      if (witness.len <= 0) {
```

Comment:
I think `<=` vs `==` is confusing, since `witness.len` is a `size_t` and cannot be below 0

Reply:
Do you think so?  I prefer `<=` to offer protection against refactorings that might change the signedness of these sort of types.  Why might the signedness change?  Because I'm told that C will somehow take advantages of UB in signed arithmetic to optimize loops, and this is something I haven't really considered yet.  Perhaps in the future I will want to change the `size_t` in loops like on line 343 to `int` and cause a ripple effect of causing preferences of `int` over `size_t` in `witness.len`.

However, if you feel strongly, please let me know, and I will change it to `==`

Reply:
No, this seems reasonable. `<=` is fine.